pgsql: extensions update (#200)
CleverCloud/documentation: content/doc/addons/postgresql/_index.md

type: docs
---

PostgreSQL is an object-relational database management system (ORDBMS) with an emphasis on extensibility and on standards-compliance.

## Versions

Versions currently available for order are as follows:

> - On our shared (DEV) cluster: PostgreSQL version 15

Reviewer: ```suggestion
- On shared (DEV) cluster: PostgreSQL version 15
```

- On dedicated plans: PostgreSQL versions 10, 11, 12, 13, 14 and 15

{{< readfile file="db-backup.md" >}}

## Migrating from an old database

Some applications require a non-empty database to run properly. If you want to import your **SQL** dump, you can use several methods:

> 1. [Our WebGUI (Adminer)](https://dbms-adminer.clever-cloud.com/)

Reviewer: ```suggestion
1. [Clever Cloud WebGUI (Adminer)](https://dbms-adminer.clever-cloud.com/)
```

Reviewer: Do we add `adminer` to the list of accepted terms? I have no idea what is it.

2. Command line tool for PostgreSQL administration like `psql`
3. Any PostgreSQL client such as [pgAdmin](https://www.pgadmin.org/)

## Direct access

> A proxy serves all dedicated PostgreSQL databases. In some cases, this can add some latency between applications and their database. If this is an issue, you must generate a direct hostname and port for the add-on to bypass the proxy, using the "Generate direct hostname and port" button in the add-on dashboard.

Reviewer 2: ```suggestion
A proxy serves all dedicated PostgreSQL databases. In some cases, this can add some latency between applications and their database. If this is an issue, you can generate a direct hostname and port for the add-on to bypass the proxy, using the "Generate direct hostname and port" button in the add-on dashboard.
```

> All our dedicated PostgreSQL databases are served through a proxy. In some cases, this can add some latency between applications and their database. If this is an issue, the proxy can be bypassed by generating a direct host-name and port for the add-on. You can do so by clicking the "Generate direct hostname and port" button in the add-on dashboard.

Reviewer: ```suggestion
A proxy serves all dedicated PostgreSQL databases. In some cases, this can add some latency between applications and their database. If this is an issue, you can bypass the proxy by generating a direct hostname and port for the add-on. Clicking the "Generate direct hostname and port" button in the add-on dashboard.
```

> Doing this will add new variables to the add-on's environment, allowing connections without going through the proxy.

Reviewer: ```suggestion
Generating direct access adds new variables to the add-on's environment, allowing connections without going through the proxy.
```

{{< callout type="info" >}} 

> Bear in mind that there is a trade-off with direct access. After migrating an add-on, you will need to re-generate the host name and port. This means applications linked through direct access will not be able to connect until you manually regenerate them. 

Reviewer: ```suggestion
Direct access doesn't persist after upgrading/migrating a database. Manually re-generate direct hostname and port after an upgrade to allow applications linked through direct access to connect to the database. 
```

Such a limitation doesn't exist when using proxy access.
{{< /callout >}}

## Note on shared databases

Encryption at rest is available on PostgreSQL. You can find more information on the [dedicated page]({{< ref "doc/administrate/encryption-at-rest.md" >}})

> This referencing does not exist for dedicated databases.

Reviewer 2: Something went horribly wrong here it seems. The entire section on the shared cluster is gone.

type: docs
---

PostgreSQL is an object-relational database management system (ORDBMS) with an emphasis on extensibility and on standards-compliance.

## Versions

The versions currently available are :

> - Shared plans (DEV) : PostgreSQL 15

Reviewer 3: ```suggestion
- Shared plans (DEV): PostgreSQL 15
```

> - Newly created dedicated databases (XS Small Space and above) : PostgreSQL 10, 11, 12, 13, 14, 15

Reviewer 3: ```suggestion
- Newly created dedicated databases (XS Small Space and above): PostgreSQL 10, 11, 12, 13, 14, 15
```

{{< readfile file="db-backup.md" >}}

## Migrating from an old database

Some applications require a populated database to run properly. If you want to import your **SQL** dump, you can use several methods:

1. [Our WebGUI (Adminer)](https://dbms-adminer.clever-cloud.com/).
2. Command line tool for PostgreSQL administration like `psql`.
3. Any PostgreSQL client such as [pgAdmin](https://www.pgadmin.org/).

## Direct access

All our dedicated PostgreSQL databases are served via a proxy. To reduce the latency you can bypass this proxy by generating direct hostname and port for the add-on. You can do it by clicking the "Generate direct hostname and port" in the add-on dashboard.

This action will add new environment variables to reach the add-on without any proxy.

> Keep in mind that there's a trade-off with direct access: during a migration, you will need to generate hostname and port again, so your application must update them. Such a modification is not necessary with proxy access.

Reviewer 2: ```suggestion
{{< callout type="warning" >}} 
Please bear in mind that there is a trade-off with direct access. After migrating an add-on, you will need to re-generate the host-name and port. This means applications linked through direct access will not be able to connect until you manually regenerate them. 

Such a limitation doesn't exist when using proxy access.
{{< /callout >}}
```

## Encryption at rest

> Encryption at rest is available on PostgreSQL. You can have more information on the [dedicated page]({{< ref "doc/administrate/encryption-at-rest.md" >}})

Reviewer 2: ```suggestion
Encryption at rest is available on PostgreSQL. You can find more information on the [relevant page]({{< ref "doc/administrate/encryption-at-rest.md" >}})
```

1. [WebGUI (Adminer)](https://dbms-adminer.clever-cloud.com/)
2. Command line tool for PostgreSQL administration like `psql`
3. Any PostgreSQL client such as [pgAdmin](https://www.pgadmin.org/)

## Direct access

A proxy serves all dedicated PostgreSQL databases. In some cases, this can add some latency between applications and their database. If this is an issue, you can generate a direct hostname and port for the add-on to bypass the proxy, using the "Generate direct hostname and port" button in the add-on dashboard.

Generating direct access adds new variables to the add-on's environment, allowing connections without going through the proxy.

{{< callout type="info" >}} 
Direct access doesn't persist after migrating a database. Manually re-generate direct host name and port after a migration to allow applications linked through direct access to connect to the database. 

Such a limitation doesn't exist when using proxy access.
{{< /callout >}}

## Encryption at rest

Encryption at rest is available on PostgreSQL. You can find more information on the [dedicated page]({{< ref "doc/administrate/encryption-at-rest.md" >}})

## Note on shared databases

> Free databases are instantiated from a shared cluster. The way postgresql works on shared clusters is that a `pg_database` table references the name, and only the name, of other databases in that same cluster. The data, on the other hand, is not accessible.

Reviewer 2: ```suggestion
If you try to list the databases on the shared cluster, you will see the names of the databases of all other users of the cluster. This is normal behaviour. Rest assured, permissions are set properly, meaning no one but your user can read or write data to your database.
```

@aurrelhebert, is this still the case? I have a feeling it was patched recently :thinking:

Author: I let you decide on this and tell me if it's ok for commit/merge or not

Reviewer 3: Not released yet ;)

Reviewer 2: Okay, then @davlgd you can apply the suggestion and then I'll approve :)

Reviewer 2: Okay, then @davlgd you can apply the suggestion and then I'll approve :)

 pgcrypto               | Cryptographic functions
 pgrowlocks             | Show row-level locking information
 pgstattuple            | Show tuple-level statistics
 plcoffee               | PL/CoffeeScript (v8) trusted procedural language
 plls                   | PL/LiveScript (v8) trusted procedural language
 plpgsql                | PL/pgSQL procedural language
 plv8                   | PL/JavaScript (v8) trusted procedural language
 postgis                | PostGIS geometry and geography spatial types and functions
 postgis_raster         | PostGIS raster types and functions
 postgis_tiger_geocoder | PostGIS tiger geocoder and reverse geocoder
 postgis_topology       | PostGIS topology spatial types and functions
 postgres_fdw           | Foreign-data wrapper for remote PostgreSQL servers
 refint                 | Functions for implementing referential integrity (obsolete)
 seg                    | Data type for representing line segments or floating-point intervals
 sslinfo                | Information about SSL certificates
 tablefunc              | Functions that manipulate whole tables, including crosstab
 tcn                    | Triggered change notifications

>  timescaledb            | Enables scalable inserts and complex queries for time-series data (Community Edition)

Reviewer 2: Timescaledb is not installed by default.

## `pg_activity`

> If you want to use [pg_activity](https://github.com/dalibo/pg_activity) on an PostgreSQL add-on, but you get `Exception: Must be run with database superuser privileges.`, you need to add the `--rds` flag when you start it.  

Reviewer 2: ```suggestion
If you want to use [pg_activity](https://github.com/dalibo/pg_activity) on a PostgreSQL add-on, but you encounter the following error `Exception: Must be run with database superuser privileges.`, you need to add the `--rds` flag when you start it.
```
